Freshness: C# fundamentals and programming guides (#49129)
dotnet/docs: docs/csharp/fundamentals/types/interfaces.md

---
title: "Interfaces - define behavior for multiple types"
description: An interface in C# contains definitions for a group of related functionalities that a non-abstract class or a struct must implement. It specifies the members and their signatures for a type that implements the interface.
ms.date: 10/13/2025
helpviewer_keywords: 
  - "interfaces [C#]"
  - "C# language, interfaces"
---
# Interfaces - define behavior for multiple types

> An interface contains definitions for a group of related functionalities that a non-abstract [`class`](../../language-reference/keywords/class.md) or a [`struct`](../../language-reference/builtin-types/struct.md) must implement. An interface might define `static` methods, which must have an implementation. An interface might define a default implementation for members. An interface might not declare instance data such as fields, automatically implemented properties, or property-like events.

Reviewer: This changed in a new version. Interfaces can now declare `abstract static` and `virtual static` members. Also, a couple might-> can (or can't) changes

```suggestion
An interface contains definitions for a group of related functionalities that a non-abstract [`class`](../../language-reference/keywords/class.md) or a [`struct`](../../language-reference/builtin-types/struct.md) must implement. An interface can define `static` methods. An interface can define a default implementation for members. An interface can't declare instance data such as fields, automatically implemented properties, or property-like events.
```

For more information about abstract classes, see [Abstract and Sealed Classes and Class Members](../../programming-guide/classes-and-structs/abstract-and-sealed-classes-and-class-members.md).

> Interfaces can contain instance methods, properties, events, indexers, or any combination of those four member types. Interfaces might contain static constructors, fields, constants, or operators. Beginning with C# 11, interface members that aren't fields might be `static abstract`. An interface can't contain instance fields, instance constructors, or finalizers. Interface members are public by default, and you can explicitly specify accessibility modifiers, such as `public`, `protected`, `internal`, `private`, `protected internal`, or `private protected`. A `private` member must have a default implementation.

Reviewer: Again, I'd go with "can" because the restriction was removed:

```suggestion
Interfaces can contain instance methods, properties, events, indexers, or any combination of those four member types. Interfaces can contain static constructors, fields, constants, or operators. Beginning with C# 11, interface members that aren't fields can be `static abstract`. An interface can't contain instance fields, instance constructors, or finalizers. Interface members are public by default, and you can explicitly specify accessibility modifiers, such as `public`, `protected`, `internal`, `private`, `protected internal`, or `private protected`. A `private` member must have a default implementation.
```
